bindings/rust/s2n-tls/src/callbacks.rs

```diff
@@ -118,30 +124,126 @@ pub(crate) fn trigger_async_callback<T: 'static + AsyncCallback>(
 /// and calling any "mark done" style methods necessary to unblock the
 /// connection once the callback has succeeded.
 pub(crate) trait AsyncCallback {
-    fn poll(&mut self, conn: &mut Connection) -> Poll<Result<(), Error>>;
+    fn poll(
+        // this should NEVER be a `&mut` since its would be possible to
+        // mutate the Config on the Connection. Since one Config can be
+        // used for multiple Connections that would be undefined-behavior
+        &self,
```

Comment:
What if there was context knowingly shared by multiple Connections? Like what if the ClientHelloCallback had a Connection->Context mapping stored, and each Connection just updated its own context?